pydantic/_internal/_generate_schema.py

```diff
@@ -766,3 +771,137 @@ def get_model_self_schema(cls: type[BaseModel]) -> core_schema.ModelSchema:
         core_schema.definition_reference_schema(model_ref),
         metadata=build_metadata_dict(js_metadata=model_js_metadata),
     )
+
+
+def apply_discriminator(schema: core_schema.CoreSchema, discriminator: str) -> core_schema.CoreSchema:
+    # Eventually: should add support for other discriminator types, and explicitly specified choices
+
+    # If the field was wrapped with a nullable schema, we preserve that here and re-wrap the final result at the end.
+    nullable_schema: core_schema.NullableSchema | None = None
+    if schema['type'] == 'nullable':
+        nullable_schema = schema
+        schema = schema['schema']
+
+    if schema['type'] != 'union':
+        raise TypeError('`discriminator` can only be used with `Union` type with more than one variant')
+
+    union_choices = schema['choices']
+    if len(union_choices) < 2:
+        raise TypeError('`discriminator` can only be used with `Union` type with more than one variant')
```

Comment:
🤷‍♂️ I think there was a test for this; either way I was just replicating the same logic that was in v1. I agree it doesn't seem there's anything wrong with a one-item union in general.

Reply:
If there's a test for it then fair to leave it as is. In general I'd prefer to handle cases like this by "doing nothing special" unless there's some really bad implications for the users.

Reply:
Not going to change this for now since, even though the error messages are the same, you'd be able to tell which error you hit in the traceback by looking at the line numbers. I think we should probably change the error message for these two lines anyway but perhaps that can come later with a copy-pass over all of our error messages.